### PR TITLE
BaseFold: Add and reimplement some utility functions.

### PR DESCRIPTION
**Pull request by yczhangsjtu**
Extracting small PRs from #294 

There was an implementation of Iterator for iterating through a `FieldType`. Turns out that it can be replaced with existing tools in `iter_tools`. This proposes the new implementation.

Also added some other utility functions for type conversion between `FieldType` and `Vec<E>`, to avoid `match field_type` everywhere.

### CODE REVIEW

**Comment by matthiasgoergens** [comment]
Thanks for breaking down the big PR into more digestible chunks!

File: mpcs/src/util.rs [hunk 1]
```diff
@@ -148,38 +148,45 @@ pub fn field_type_index_set_ext<E: ExtensionField>(
...
-            None
-        } else {
-            let res = field_type_index_ext(self.inner, self.index);
-            self.index += 1;
-            Some(res)
-        }
+pub fn field_type_to_ext_vec<E: ExtensionField>(evaluations: &FieldType<E>) -> Vec<E> {
+    match evaluations {
+        FieldType::Ext(coeffs) => coeffs.to_vec(),
+        FieldType::Base(coeffs) => coeffs.iter().map(|x| (*x).into()).collect(),
```
Comment (matthiasgoergens) [inline mpcs/src/util.rs:170]: Cosmetic suggestion:

```suggestion
        FieldType::Base(coeffs) => coeffs.iter().map(|&x| x.into()).collect(),
```

**Review by matthiasgoergens — COMMENTED** [review]
Could you please add in the PR description _why_ we are making this refactoring?

Just a quick note about what task becomes easier or simpler with the new organisation?

Thanks!

File: mpcs/src/util.rs [hunk 2]
```diff
@@ -148,38 +148,45 @@ pub fn field_type_index_set_ext<E: ExtensionField>(
...
-pub fn field_type_iter_ext<E: ExtensionField>(evaluations: &FieldType<E>) -> FieldTypeIterExt<E> {
-    FieldTypeIterExt {
-        inner: evaluations,
-        index: 0,
+pub fn field_type_iter_base<E: ExtensionField>(
+    values: &FieldType<E>,
+) -> impl Iterator<Item = &E::BaseField> + '_ {
+    match values {
+        FieldType::Ext(coeffs) => Either::Left(coeffs.iter().flat_map(|x| x.as_bases())),
+        FieldType::Base(coeffs) => Either::Right(coeffs.iter()),
```
[the remark this answers is not in the document]
Comment (yczhangsjtu) [reply mpcs/src/util.rs:188]: No the return types are not the same. They are both iterators over BaseField but different types of Iterator.

[the remark this answers is not in the document]
Comment (yczhangsjtu) [reply mpcs/src/util.rs:188]: The `Either` is some solution I found on the internet after many fruitless struggles. Any better suggestions are welcome. I'm really not an expert in rust iterators.

[the remark this answers is not in the document]
Comment (hero78119) [reply mpcs/src/util.rs:188]: Another option is rely on runtime dispatch.

```diff
diff --git a/mpcs/src/util.rs b/mpcs/src/util.rs
index 7688b53e..6edd82d4 100644
--- a/mpcs/src/util.rs
+++ b/mpcs/src/util.rs
@@ -182,10 +182,10 @@ pub fn field_type_as_ext<E: ExtensionField>(values: &FieldType<E>) -> &Vec<E> {

 pub fn field_type_iter_base<E: ExtensionField>(
     values: &FieldType<E>,
-) -> impl Iterator<Item = &E::BaseField> + '_ {
+) -> Box<dyn Iterator<Item = &E::BaseField> + '_> {
     match values {
-        FieldType::Ext(coeffs) => Either::Left(coeffs.iter().flat_map(|x| x.as_bases())),
-        FieldType::Base(coeffs) => Either::Right(coeffs.iter()),
+        FieldType::Ext(coeffs) => Box::new(coeffs.iter().flat_map(|x| x.as_bases())),
+        FieldType::Base(coeffs) => Box::new(coeffs.iter()),
         _ => unreachable!(),
     }
 }
```

After reading document in Either, I think static dispatch via Either is better, and either already implement iterator so it's truly works this purpose. So LGTM under this scope! 👍

**Review by hero78119 — APPROVED** [review]
👍

**Pull request merged**